File: coin-change/coin-change.cs
public class Solution {
    public int CoinChange(int[] coins, int amount) {
        if(amount<1) return 0;
        
        return CoinChange(coins,amount,new int[amount]);
        
    }
    
    public int CoinChange(int[] coins,int rem,int[] count)
    {
        if(rem<0) return -1;
        if(rem==0) return 0;
        
        if(count[rem-1]!=0) return count[rem-1];
        int min = Int32.MaxValue;
        foreach(int coin in coins)
        {
            int res = CoinChange(coins,rem-coin,count);
            if(res>=0 && res < min)
            {
                min = 1+res;
            }
        }
        
        count[rem-1] = (min==Int32.MaxValue) ?-1 :min;
        return count[rem-1];
    }
}

File: problem1/CoinChange.java
// Time Complexity : O(n*n) : n -> size of array
// Space Complexity : O(amount*n) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

package problem1;

import java.util.Arrays;

public class CoinChange {
	/*
	 * Dynamic Programming. We start with 1 coin, and check number of coins required
	 * to make up amounts from 0 to amount. We either choose the current coin, or
	 * not.
	 */
	public int coinChange(int[] coins, int amount) {
		if (coins == null || coins.length == 0 || amount == 0) {
			return 0;
		}

		int[][] dp = new int[coins.length + 1][amount + 1];

		Arrays.sort(coins);

		for (int i = 0; i <= coins.length; i++) {
			dp[i][0] = 0;
		}

		for (int j = 0; j <= amount; j++) {
			dp[0][j] = Integer.MAX_VALUE - 1;
		}

		for (int i = 1; i <= coins.length; i++) {
			for (int j = 1; j <= amount; j++) {
				if (coins[i - 1] > j) {
					dp[i][j] = dp[i - 1][j];
				} else {
					dp[i][j] = Math.min(dp[i - 1][j], dp[i][j - coins[i - 1]] + 1);
				}
			}
		}
		return dp[coins.length][amount] != Integer.MAX_VALUE - 1 ? dp[coins.length][amount] : -1;
	}

	public static void main(String[] args) {
		CoinChange obj = new CoinChange();
		System.out.println("Min coins required = " + obj.coinChange(new int[] { 1, 2, 5 }, 11));
	}

}
